fix: Overlap for long passwords and show/hide button in Password Auth Screen Issue #: 5614

KshitizSareen opened the pull request:
In the Password Auth Screen, when you type a long password, it overlaps with the show/hide button.

I fixed this issue by creating a new stylesheet for the password text input field. I added a padding-right property of 15% to its style. This fix will ensure that the text input in the password field will never exceed a width of 85% of the text input field.

I tested this commit by viewing the password auth screen in horizontal and vertical views. The password input field is working fine in both instances.

Please let me know if you have any questions.

Thank you.

Fixes: #5614.

chrisbobbe commented:
Hi @KshitizSareen, thank you!

Could you please post some [screenshots](https://zulip.readthedocs.io/en/latest/git/pull-requests.html) of the proposed change?

KshitizSareen commented:
Hi @chrisbobbe, Thank you for your message. Let me attach the screenshots.

KshitizSareen commented:
https://user-images.githubusercontent.com/23134343/208796664-d337a14f-d3e7-4be4-b0f9-b878d3725a6d.mov

@chrisbobbe I have attached a small video showing the fix.

Please let me know if you have need any more details.

Regards,
Kshitiz.

chrisbobbe commented on src/common/PasswordInput.js:
```diff
@@ -16,11 +16,19 @@ const styles = createStyleSheet({
     color: BRAND_COLOR,
   },
+  passwordTextInput: {
+    paddingRight: '15%',
```
Hmm, I'd rather not use a percentage here. We should reserve exactly the right amount of space for the show/hide button, and this will cause that space to be too large or small depending on factors including:
- How wide the input is (e.g., varies with screen width; will be large on a tablet)
- How big the show/hide button's text is (varies with the user's chosen font size in system settings)

chrisbobbe commented on src/common/PasswordInput.js:
```diff
@@ -16,11 +16,19 @@ const styles = createStyleSheet({
     top: 0,
     bottom: 0,
     justifyContent: 'center',
+    borderTopWidth: 1,
+    borderColor: BORDER_COLOR,
+    borderBottomWidth: 1,
+    borderRightWidth: 1,
```
I expect this to look odd on Android, where the inputs aren't styled with a border (see use of `Platform.select` in Input.js).

chrisbobbe reviewed (changes requested):
Thanks! Comments below.

I think they're most easily addressed by giving up the design where the show/hide button appears inside the border (which is iOS only). Let's try removing the absolute positioning, and just let the button take up the horizontal space it needs, with the `<Input />` stretching to fill what horizontal space is left.

Here's a sketch of a possible implementation:
- `flexDirection: 'row'` on the parent `<View />`
- `flex: 1` on the `<Input />`
- Remove the `position`, `right`, `top`, and `bottom` attributes of `styles.showPasswordButton`.

I think it makes sense for this change to be in just one [clear, coherent commit](https://github.com/zulip/zulip-mobile/blob/main/docs/style.md#git-commits-commit-messages).

KshitizSareen commented:
Hi @chrisbobbe ,

Thank you so much for your comments. I tried to incorporate your changes. 

I removed the absolute positioning and added flex-direction: 'row' to the view. Unfortunaltely, that wont look very good with borders. It looks something like this:

<img width="306" alt="Screenshot 2022-12-28 at 3 18 01 PM" src="https://user-images.githubusercontent.com/23134343/209884091-f8eaf14c-0475-443f-955c-992fde1ad4f6.png">

We could do away with borders entirely, to make it look like this (Similar to Android):

<img width="314" alt="Screenshot 2022-12-28 at 3 19 30 PM" src="https://user-images.githubusercontent.com/23134343/209884174-7cfc8d8f-233a-4c5e-965c-19c0710a3ea0.png">

Another issue is the margin of the show/hide button. Since show has a bigger width than height, its margin causes the password text input to shrink. It works something like this:


https://user-images.githubusercontent.com/23134343/209884399-0fa06b22-be59-4e94-a5e4-ace83c33bc7c.mov

To resolve this issue, I propose we could use some icons instead of the show/hide text. It looks better and more intuitive.
We could use icons like these instead of show/hide:

<img width="341" alt="Screenshot 2022-12-28 at 3 26 54 PM" src="https://user-images.githubusercontent.com/23134343/209884554-77874623-6df0-4ed8-9324-35d01c88c2d5.png">
<img width="375" alt="Screenshot 2022-12-28 at 3 27 00 PM" src="https://user-images.githubusercontent.com/23134343/209884560-1a40a09b-8c4d-4418-bbdd-5704794f4f52.png">

Since they always occupy a fixed size, their margin also wont cause a problem.

Please let me know if you want to discuss on this. I look forward to your response.

Thank you,
Regards,
Kshitiz.

chrisbobbe commented:
> To resolve this issue, I propose we could use some icons instead of the show/hide text. It looks better and more intuitive.
We could use icons like these instead of show/hide:

Sure, that seems like a fine thing to try. Do you have the development environment set up to test on Android too? We style the `Input` component differently on Android and iOS, and we'd like to make sure it looks reasonable on both. If it's difficult for you to set up, please let me know and I can test your revision and post screenshots.

KshitizSareen commented:
Hi @chrisbobbe, Thanks for your comments. I have emulators for the platforms, and I will make sure to test on both Android and IOS. Let me code it up and show the proposed changes.

Thanks.

richab246 commented:
Can I work on this issue?

chrisbobbe commented:
@rishab246, please find a different issue to work on; @KshitizSareen is working on this one.

KshitizSareen commented:
Hi @chrisbobbe, I made the proposed changes. Please review them.

This is for Android:
https://user-images.githubusercontent.com/23134343/211293421-98e2d07e-6e6b-4ffa-be58-f6676f8ea2d9.mov

This is for IOS:
https://user-images.githubusercontent.com/23134343/211294324-dadbbbb0-d3e8-41f8-bdae-f6f16c13d947.mov

Thank you,
Regards,
Kshitiz.

chrisbobbe commented on src/common/PasswordInput.js:
```diff
@@ -44,10 +46,16 @@ export default function PasswordInput(props: Props): Node {
+      />
       <Touchable style={styles.showPasswordButton} onPress={handleShow}>
-        <ZulipTextIntl style={styles.showPasswordButtonText} text={isHidden ? 'show' : 'hide'} />
+        <Icon name={isHidden ? 'eye-off' : 'eye'} size={20} style={styles.showPasswordButtonIcon} />
```
We default to using size 24 for icons; what does it look like with 24 instead of 20?

chrisbobbe commented on src/common/PasswordInput.js:
```diff
@@ -3,24 +3,26 @@ import React, { useState, useCallback } from 'react';
 import type { Node } from 'react';
 import { View } from 'react-native';
 
+import Icon from 'react-native-vector-icons/Ionicons';
```
```suggestion
import { Icon } from '../common/Icons';
```

chrisbobbe reviewed (changes requested):
Thanks, @KshitizSareen! See two small comments below. Also, in your next revision, please format your commit message according to our style (I linked to it upthread).

KshitizSareen commented:
Hi @chrisbobbe,

Thanks for your comments. I changed the font-size to 24.

It looks like this:

This is for tablet:
https://user-images.githubusercontent.com/23134343/213664807-cdd6c6ee-1adc-428f-8c1f-0268c1707c1a.mov

This is for Iphone:

https://user-images.githubusercontent.com/23134343/213664902-f21c406d-f33b-496e-a2cf-2754b54149d8.mov

If you thinks it fine, I will push the changes.

Thank you.

chrisbobbe commented:
Thanks! Please squash your fixup commits into the main, interesting commit, and give that commit a commit message according to our style.

KshitizSareen commented:
Hi @chrisbobbe , I squashed all my commits.

Please review.

Thanks.

chrisbobbe commented:
Thanks! Some commit-message nits before I hand this off to Greg to review:

- Please make sure your summary line is 76 characters or less.
- For the prefix, use the name of a subsystem instead of "fix". The name of a UI component often works; I think "PasswordInput" would be fine here.
- Please end your commit message with `Fixes: #5614` (no period, capital "F").

This is all covered in https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html. I highly recommend [browsing the project's history](https://github.com/zulip/zulip-mobile/blob/main/docs/howto/git.md#reading-history) to get a sense of our commit-message style, in addition to reading the docs.

KshitizSareen commented:
Thank you @chrisbobbe.

I will change the commit message, and I will be more careful from now on.

Thank you.

KshitizSareen commented:
Hi @chrisbobbe , I pushed some changes.

Please review.

Thanks.

chrisbobbe commented:
Thanks! A few nits in the updated commit message:

```
PasswordInput: Fixed Overlap for long Passwords and show/hide button

Long Passwords and show/hide buttons overlap with each other.
I fixed this problem by doing the following:

- I changed the style of Password Text Input to take up as much space as needed. I did this by setting a flex of 1 to the password text input.

- I changed the positioning of the show/hide button from absolute to relative.

- I ensured Horizontal Alignment by setting the flex-direction of the parent view to 'row'.

- To fix the problem of the changing width of the text input, when the show/hide text changes, I replaced the show/hide text with a show/icon. The show/hide icon is an eye-open/eye-closed icon. This icon always has a fixed width and does not affect the text input.

- I removed hide text from messages_en.json since it is not in the app.

I tested this on Android and IOS devices with different screen sizes, and it works fine on both platforms.

Fixes: #5614
```

- The summary line should say "Fix" instead of "Fixed", and there's no need to capitalize "overlap" or "passwords".

- The paragraph content in the body should be wrapped to about 68 characters per line, but no more than 70.

https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html#commit-messages

chrisbobbe commented:
Thanks! I've just pushed a revision with a few small tweaks to the layout; please take a look:

c0d475a5e PasswordInput: Fix overlap for long passwords and show/hide button
4b7887524 PasswordInput: Add 8px margin between password input and show/hide button
3682bc313 PasswordInput: Don't let show/hide button stretch password input's height
322b330a0 PasswordInput: Give show/hide button a 48px touch target size
3f3324fef PasswordInput: Use Pressable for show/hide icon button
9c473e1f3 PasswordInput: Add accessibility properties to show/hide button

Here's a video from after those small changes, to compare with the video of your revision that you posted (thanks again 🙂) at https://github.com/zulip/zulip-mobile/pull/5621#issuecomment-1398144516:

<img width=400 src="https://user-images.githubusercontent.com/22248748/218187701-77ecc628-0403-46f2-af7d-c2080e51f6a6.gif" />

@gnprice, this is ready for your review. 🙂
